include/crm/common/xml_internal.h

```diff
@@ -310,4 +310,46 @@ pcmk__xe_first_attr(const xmlNode *xe)
 char *
 pcmk__xpath_node_id(const char *xpath, const char *node);
 
+/*!
+ * \internal
+ * \brief Add a boolean attribute to an XML node.
+ *
+ * \param[in,out] node  XML node to add attributes to
+ * \param[in]     name  XML attribute to create
+ * \param[in]     value Value to give to the attribute
+ */
+void
+pcmk__xe_set_bool_attr(xmlNodePtr node, const char *name, bool value);
+
+/*!
+ * \internal
+ * \brief Extract a boolean attribute's value from an XML element
+ *
+ * \param[in] node XML node to get attribute from
+ * \param[in] name XML attribute to get
+ *
+ * \return True if the given \p is an attribute on \p node and has the value
+ *         "true", False in all other cases
+ */
+bool
+pcmk__xe_get_bool_attr(xmlNodePtr node, const char *name);
+
+/*!
+ * \internal
+ * \brief Determine whether a boolean attribute's value is False or not
+ *
+ * \param[in] node XML node to get attribute from
+ * \param[in] name XML attribute to get
+ *
+ * \return True if the given \p is an attribute on \p node and has the value
+ *         "false", False in all other cases.
+ *
+ * \note Because pcmk__xe_get_bool_attr can return False if \p name does not
+ *       exist, there is no reliable way to use it to verify that an attribute
+ *       both exists and has the False value.  This function allows for testing
+ *       that one specific case.
+ */
+bool
+pcmk__xe_bool_attr_is_false(xmlNodePtr node, const char *name);
```

Comment:
Alternative approaches using a single getter:

    /* Return standard code (EINVAL for NULL args, ENODATA for no value,
       pcmk_rc_unknown_format for invalid value), with value as output argument */
    int pcmk__xe_get_bool_attr(xmlNodePtr node, const char *name, bool *value);

    // Return default_value if not present or invalid
    bool pcmk__xe_get_bool_attr(xmlNodePtr node, const char *name, bool default_value);

The first would handle the remaining complex cases, and would parallel the crm_element_value_TYPE() functions better, but the need for a variable is annoying if you just want one quick check. It might be worthwhile to implement along with a wrapper for the common case -- back to two functions but more flexible.

Reply:
I think the first is probably the better way to go, but there's still a couple complicated cases where we're not going to be able to use these style functions regardless.  I'll try to point them out in a future comment.

Reply:
Okay, the few outstanding places that could potentially be changed are in get_ordering_symmetry and get_minimum_first_instances.  I think the former could benefit from a function like crm_element_value_default, which could also probably be more widely used.